lib/s3.js

```diff
@@ -141,4 +140,30 @@ module.exports = CoreObject.extend({
       });
     });
   },
+
+  listAllObjects: function(options) {
+    var client         = this._client;
+    var listObjects    = Promise.denodeify(client.listObjects.bind(client));
+    var allRevisions   = [];
+
+    return new Promise(function(resolve, reject) {
```

Comment:
Rather than low-level function style, could this `return` promises, which wait naturally?

```js
function listObjectRecursively(options) {
  return listObjects(options).then(function(response) {
    ...
    if (response.isTruncated) {
      ...
      return listObjectRecursively(options);
    } else {
      ...
      return response;
    }
  });
}

return listObjectRecursively(options);
```